### PR TITLE
Flexible Dedicated Hosting Options for Mission Files

### PR DESCRIPTION
**Pull request by justin-lovell**
## Background

Hello! When I was looking over the options available for dedicated server hosting, the biggest constraint was around how the file systems needed to mirror exactly on the generator and server.

Furthermore, when playing as a dedicated server, the mission may restart and wipe out a whole squadrons mission complete on reboot.

Server administrators may not want to allow access to the servers either which makes game masters ("organizing") a tad difficult.

## Solution

This pull request would solve the various problems above by reversing some prior logic to enable the above problems to be solveable. 

* Administrators are able to direct state files to specific directories, and use other tools like Dropbox to send back the state to the game master.
* Each state file is versioned to the start of the mission, just in case.

These are controlled by two environment variables.
* `LIBERATION_EXPORT_DIR` would determine where the directory is located (e.g. a Dropbox location, FTP, whatever)
* `LIBERATION_EXPORT_STAMPED_STATE` would stamp each mission start-up to ensure that server restarts and human errors are not felt across the webby-squadrons!

If these variables are not set, the Liberation and DCS work cycle remains as-is today.

## Other notes

Most of the code was in place, and it just required minor corrections. I've also refactored the code a bit to isolate it a bit of a unit.

### CODE REVIEW

File: resources/scripts/dcs_liberation.lua [hunk 1]
```diff
@@ -39,46 +39,54 @@ write_state = function()
...
+        if not (directoryOrFilePath:sub(-#filename) == filename) then
+            return directoryOrFilePath .. filename
         end
+
+        return directoryOrFilePath
     end
-end
-if debriefing_file_location then
-    local filename = "state.json"
-    if not debriefing_file_location:sub(-#filename) == filename then
```
Comment (justin-lovell) [inline resources/scripts/dcs_liberation.lua:74]: It is noteworthy that this line would have never worked. This has been corrected with bracing at the `not` would operate on the `debriefing_file_location:sub(-#filename)` statement, and not on the expanded binary statement of `debriefing_file_location:sub(-#filename) == filename`

E.g. Original code is interpretted as
```
-- original
if (not debriefing_file_location:sub(-#filename)) == filename then
-- ^ --- note this bracket for emphasis ------- ^

-- corrected in PR
if not (debriefing_file_location:sub(-#filename) == filename) then
--     ^ --- note this bracket for emphasis --------------- ^
```

**Comment by Khopa** [comment]
Hi Justin, 

That's a good idea. 
Thanks for fixing and improving that !

**Comment by davidp57** [comment]
Is that me or did you remove the export to %TEMP% ?
In your description this is your 3rd option :
```
    -- establish a search pattern into the following modes
    -- 1. Environment variable mode, to support dedicated server hosting
    -- 2. Embedded DCS Liberation Generation, to support locally hosted single player
    -- 3. Retain the classic TEMP directory logic
```
But I can't find it anywhere.

**Comment by justin-lovell** [comment]
@davidp57 , my bad - my eye must have skipped it during the self-review. I have addressed it here: https://github.com/Khopa/dcs_liberation/pull/231


Thanks for spotting it - definitely wanted to retain the original intent.

**Pull request merged**